### PR TITLE
Fix overwritting downloaded files with same name

### PR DESCRIPTION
**Pull request by gtam25**
Added auto-numbering for downloaded file names, which conflict with file names in the current directory. #6068 

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)

### CODE REVIEW

File: src/main/java/org/jabref/logic/util/io/FileNameHandler.java [hunk 1]
```diff
@@ -67,7 +71,41 @@ public static String cleanDirectoryName(String badFileName) {
...
+     * @return a file-name such that it does not match any existing files in targetDirectory.
+     * */
+    public static String getNonOverWritingFileName(Path targetDirectory, String fileName) {
+        String absoluteName = targetDirectory.resolve(fileName)
+                                            .toString();
+
+        Optional<String> extensionOptional = FileUtil.getFileExtension(fileName);
+
+        // the suffix include the '.' , if extension is present Eg: ".pdf"
+        String extensionSuffix = "";
```
Comment (koppor) [inline src/main/java/org/jabref/logic/util/io/FileNameHandler.java]: Please put that into an else path in the if of line 91. (Declaring the variables here is OK). I find it hard if the contents of the variables are overwritten: The default case is that there is an extension.

File: src/main/java/org/jabref/logic/util/io/FileNameHandler.java [hunk 2]
```diff
@@ -67,7 +71,41 @@ public static String cleanDirectoryName(String badFileName) {
...
+
+        if (extensionOptional.isPresent()) {
+            extensionSuffix = '.' + extensionOptional.get();
+            fileNameWithoutExtension = fileName.substring(0, fileName.lastIndexOf('.'));
+        }
+
+        Path absolutePath = Paths.get(absoluteName);
+        String newFileName = fileName;
+
+        for (int counter = 1; Files.exists(absolutePath); ++counter) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/util/io/FileNameHandler.java]:
```suggestion
        for (int counter = 1; Files.exists(absolutePath); counter++) {
```

`++` at the end to be inline with the usual `i++`.

Comment (koppor) [reply src/main/java/org/jabref/logic/util/io/FileNameHandler.java]: After thinking longer, please rewrite to a `while` loop. Not testing for the counter in the test of a `for` loop is uncommon. Thus, hard to read when maintaining the code.

File: src/main/java/org/jabref/logic/util/io/FileNameHandler.java [hunk 3]
```diff
@@ -67,7 +71,41 @@ public static String cleanDirectoryName(String badFileName) {
...
+            newFileName = fileNameWithoutExtension +
+                            " (" + counter + ")" +
+                            extensionSuffix;
+            absolutePath = targetDirectory.resolve(newFileName);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/util/io/FileNameHandler.java]: No need for that variable. Can be tesed in the while condition.

File: src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    @Test
+    public void testCleanCurlyBracesAsWell() {
+        assertEquals("The Evolution of Sentiment_ Analysis_A Review of Research Topics, Venues, and Top Cited Papers.PDF", FileNameHandler.cleanFileName("The Evolution of Sentiment} Analysis}A Review of Research Topics, Venues, and Top Cited Papers.PDF"));
+    }
+
+    @Test
+    public void testGetNonOverWritingFileNameReturnsUniqueName(@TempDir Path tempDirectory) throws IOException {
+        String dummyFile1 = "default.txt";
+        String dummyFile2 = "default (1).txt";
+        String expectedFileName = "default (2).txt";
```
Comment (koppor) [inline src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java]: Please include that in the `assertEquals` directly.

File: src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    public void testCleanCurlyBracesAsWell() {
+        assertEquals("The Evolution of Sentiment_ Analysis_A Review of Research Topics, Venues, and Top Cited Papers.PDF", FileNameHandler.cleanFileName("The Evolution of Sentiment} Analysis}A Review of Research Topics, Venues, and Top Cited Papers.PDF"));
+    }
+
+    @Test
+    public void testGetNonOverWritingFileNameReturnsUniqueName(@TempDir Path tempDirectory) throws IOException {
+        String dummyFile1 = "default.txt";
```
Comment (koppor) [inline src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java]: Please introduce `Path dummyFile1` variable. In both usages of `dummyFileX`, you are resulting the name using the same path.

File: src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        assertEquals("illegalFilename\\_____.txt", FileNameHandler.cleanDirectoryName("illegalFilename\\?*<>|.txt"));
+    }
+
+    @Test
+    public void testCleanCurlyBracesAsWell() {
+        assertEquals("The Evolution of Sentiment_ Analysis_A Review of Research Topics, Venues, and Top Cited Papers.PDF", FileNameHandler.cleanFileName("The Evolution of Sentiment} Analysis}A Review of Research Topics, Venues, and Top Cited Papers.PDF"));
+    }
+
+    @Test
+    public void testGetNonOverWritingFileNameReturnsUniqueName(@TempDir Path tempDirectory) throws IOException {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/util/io/FileNameHandlerTest.java]: Good to have a test. - Could you add two more:

1. One for a non existing file
2. One for just one file existing

I know, these are trivial. - Nevertheless, if someone touches your code, even the trivial ones are tested then (and there could something go wrong)

**Comment by gtam25** [comment]
> Thank you for your contribution.
> 
> Some nitpick comments inside 😇
> 
> Two more things:
> 
> 1. Please keep `FileNameCleaner` and introduce `FileNameUniqueness`. Reason: There is no shared code between the two utility methods (no [cohesion](https://en.wikipedia.org/wiki/Cohesion_(computer_science)) in the sence of data; only in the sense of the general setting). Small classes make it easier to maintain them.
> 2. Can you try to include a check of `Files.mismatch` (see https://stackoverflow.com/a/56680991/873282 for details) at download -> before the same file is linked, just do not link it instead of creating `... (1).pdf`.

I think our problem is that different files are getting downloaded as "default.pdf" (when no BibTeX key exists).
If we use `Files.mismatch` (as you mentioned in number 2 above), since both paths refer to same file path, it would return a `match` and hence not download the other files (which we do not want I guess ?)

**Comment by koppor** [comment]
> > 2. Can you try to include a check of `Files.mismatch` (see [stackoverflow.com/a/56680991/873282](https://stackoverflow.com/a/56680991/873282) for details) at download -> before the same file is linked, just do not link it instead of creating `... (1).pdf`.
> I think our problem is that different files are getting downloaded as "default.pdf" (when no BibTeX key exists).

Yeah.

> If we use `Files.mismatch` (as you mentioned in number 2 above), since both paths refer to same file path, it would return a `match` and hence not download the other files (which we do not want I guess ?)

Long solution would be:

1. `default.pdf` exists
2. Download as `default (1).pdf`.
3. Store it in the linked files
4. Check if `default.pdf` equals `default (1).pdf`.
5. If `default.pdf` equals `default (1).pdf` then a) remote `default (1).pdf` from the file system, b) remove `default (1).pdf` from the list of linked files, c) add a notification that the file already exists.

I thought that one can omit step 3. - step 5c is necessary in all cases.

I see that this is not strongly part of your PR, so can be left for future work.

File: src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java [hunk 7]
```diff
@@ -38,6 +38,7 @@
 import org.jabref.logic.externalfiles.LinkedFileHandler;
 import org.jabref.logic.l10n.Localization;
 import org.jabref.logic.net.URLDownload;
```
Comment (eugenevasin) [inline src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java]: Hello world

Comment (eugenevasin) [reply src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java]: Hey

File: src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java [hunk 8]
```diff
@@ -38,6 +38,7 @@
 import org.jabref.logic.externalfiles.LinkedFileHandler;
```
Comment (eugenevasin) [inline src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java]: Hello

File: CHANGELOG.md [hunk 9]
```diff
@@ -25,7 +25,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 - We fixed an issue where opening a library from the recent libraries menu was not possible. [#5939](https://github.com/JabRef/jabref/issues/5939)
 - We fixed an issue with inconsistent capitalization of file extensions when downloading files [#6115](https://github.com/JabRef/jabref/issues/6115)
```
Comment (eugenevasin) [inline CHANGELOG.md]: Hello world
приветик

File: CHANGELOG.md [hunk 10]
```diff
@@ -25,7 +25,7 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
...
 - We fixed the display of language and encoding in the preferences dialog. [#6130](https://github.com/JabRef/jabref/pull/6130)
-
+- We fixed an issue where search full-text documents downloaded files with same name, overwriting existing files. [#6174](https://github.com/JabRef/jabref/pull/6174)
 ### Removed
```
Comment (eugenevasin) [inline CHANGELOG.md]: Hello world

Comment (eugenevasin) [reply CHANGELOG.md]: Hello!

File: src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Returns an absolute path to a file which does not match with any existing file names
+     *
+     * @param targetDirectory The directory in which file name should be unique
+     * @param fileName Suggested name for the file
+     * @return a file-name such that it does not match any existing files in targetDirectory.
+     * */
+    public static String getNonOverWritingFileName(Path targetDirectory, String fileName) {
+//        String absoluteName = targetDirectory.resolve(fileName)
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java]: Please remove this comment.

File: src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        if (extensionOptional.isPresent()) {
+            extensionSuffix = '.' + extensionOptional.get();
+            fileNameWithoutExtension = fileName.substring(0, fileName.lastIndexOf('.'));
+        }
+        else {
+            extensionSuffix = "";
+            fileNameWithoutExtension = fileName;
+        }
+
+//        Path absolutePath = Paths.get(absoluteName);
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java]: This one as well.

File: src/test/java/org/jabref/logic/util/io/FileNameUniquenessTest.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+public class FileNameUniquenessTest {
+
+    @TempDir
+    protected Path tempDir;
+
+    @Test
+    public void testGetNonOverWritingFileNameReturnsSameName(@TempDir Path tempDirectory) throws IOException {
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/util/io/FileNameUniquenessTest.java]: Is there a need to have the temporary directory as a variable as well as an argument? I would prefer if all tests use the same format (either variable or either paramater)

File: src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
+package org.jabref.logic.util.io;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+public class FileNameUniqueness {
+
+    /**
+     * Returns an absolute path to a file which does not match with any existing file names
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/util/io/FileNameUniqueness.java]: The description seems to be off as no absolute path is returned but only the file name, right?

File: src/test/java/org/jabref/logic/util/io/FileNameUniquenessTest.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    @Test
+    public void testGetNonOverWritingFileNameReturnsUniqueNameOver1Conflict() throws IOException {
+        Path dummyFilePath1 = tempDir.resolve("differentFile.txt");
+
+        Files.createFile(dummyFilePath1);
+
+        String outputFileName = FileNameUniqueness.getNonOverWritingFileName(tempDir, "differentFile.txt");
+        assertEquals("differentFile (1).txt", outputFileName);
+
+        Files.delete(dummyFilePath1);
```
Comment (tobiasdiez) [inline src/test/java/org/jabref/logic/util/io/FileNameUniquenessTest.java]: Deleting the file shouldn't be necessary as the `TempDir` should take care of this.

**Comment by koppor** [comment]
Comments addressed. There are some code style issues, which I will quickly fix in `master` after merge.

**Comment by koppor** [comment]
Thank you for having kept up the work. Looking forward to more contributions!

**Pull request merged**